Name-Nisarg Patel
Student ID-8878184

In [1]:
# Importing the necessary libraries

import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline 
from sklearn.metrics import mean_absolute_error , mean_absolute_percentage_error , r2_score

In [3]:
X, y = datasets.load_diabetes(return_X_y=True, as_frame=True) 
# returns 2 dataframes 
# 1. X(indepedent values) 
# 2. y(depedent value)

In [4]:
X.head(10)

age       sex       bmi        bp        s1        s2        s3  \
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
5 -0.092695 -0.044642 -0.040696 -0.019442 -0.068991 -0.079288  0.041277   
6 -0.045472  0.050680 -0.047163 -0.015999 -0.040096 -0.024800  0.000779   
7  0.063504  0.050680 -0.001895  0.066630  0.090620  0.108914  0.022869   
8  0.041708  0.050680  0.061696 -0.040099 -0.013953  0.006202 -0.028674   
9 -0.070900 -0.044642  0.039062 -0.033214 -0.012577 -0.034508 -0.024993   

         s4        s5        s6  
0 -0.002592  0.019908 -0.017646  
1 -0.039493 -0.068330 -0.092204  
2 -0.002592  0.002864 -0.025930  
3  0.034309  0.022692 -0.009362  
4 -0.002592 -0.031991 -0.046641  
5 -0.076395 -0.041180 -0.096346  
6 -0.039493 -0.062913 -0.038357  
7  0.017703 -0.035817  0.003064  
8 -0.002592 -0.014956  0.011349  
9 -0.002592  0.067736 -0.013504

In [5]:
y.head()

0    151.0
1     75.0
2    141.0
3    206.0
4    135.0
Name: target, dtype: float64

In [6]:
X.shape , y.shape

((442, 10), (442,))

In [7]:
#Get the data, and run a train-validation-test split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=42) 

In [8]:
#Run a multivariate linear regression on all variables

model_all = LinearRegression()
model_all.fit(X_train,y_train)

LinearRegression()

In [9]:
# Make predictions using the testing set
y_pred_all = model_all.predict(X_val)
y_pred_train_all = model_all.predict(X_train)

# The coefficients and intercept
print('Coefficients or the Slope(m) of the model is: ')
cols = X.columns
equ =  f"y = {model_all.intercept_:.2f}"
for i,j in  enumerate(model_all.coef_):
    print(f'{cols[i]}\t {j}')
    equ += f'+ {j:.2f}x^{i+1}'
print('intercept(c) of the model is: \t',model_all.intercept_)
print('Equation: ',equ)

Coefficients or the Slope(m) of the model is: 
age	 7.035616719487141
sex	 -246.78499668983326
bmi	 547.2491789384372
bp	 324.7966481687193
s1	 -921.8551151308552
s2	 472.11248599291594
s3	 172.47091173937798
s4	 292.8510752140103
s5	 764.5933899283848
s6	 83.23871305648963
intercept(c) of the model is: 	 147.8365248135677
Equation:  y = 147.84+ 7.04x^1+ -246.78x^2+ 547.25x^3+ 324.80x^4+ -921.86x^5+ 472.11x^6+ 172.47x^7+ 292.85x^8+ 764.59x^9+ 83.24x^10


In [10]:
# Making function for building a pipeline
def makeModelPipe(X_data,y_data):
    model = Pipeline([('polynomial', PolynomialFeatures(degree=2 , include_bias=False)), 
                    ('linear', LinearRegression())])
    return model.fit(X_data, y_data)

# The coefficients and intercept
def getEqu(model):
    print('Coefficients or the Slope(m) of the model is: ')
    equ =  f"y = {model[1].intercept_:.2f}"
    for i,j in  enumerate(model[1].coef_[1:]):
        print(f'{i}\t {j}')
        equ += f'+ {j:.2f}x^{i+1}'
    print('intercept(c) of the model is: \t',model[1].intercept_)
    print('Equation: ',equ)
    
# making function for visualization
def plot_model(X, y, model):
    # Calculate the number of rows and columns for the subplots
    X_lin = np.linspace(min(X), max(X), 100)
    plt.figure(figsize=[15,7])
    plt.title('')
    plt.scatter(X, y, s=20, label="data")
    plt.plot(X_lin, model.predict(X_lin),color='r', label="Model", lw=3)
    plt.xlabel("data")
    plt.ylabel("target")
    plt.legend()
    plt.show()

In [11]:
#Run a polynomial regression of the 2nd degree on the BMI feature alone

# Getting only the bmi data
X_train_bmi = np.array(X_train['bmi']).reshape(-1,1)
X_val_bmi = np.array(X_val['bmi']).reshape(-1,1)

# Make predictions using the testing set
model_bmi = makeModelPipe(X_train_bmi,y_train)
y_pred_bmi = model_bmi.predict(X_val_bmi)
y_pred_train_bmi = model_bmi.predict(X_train_bmi)

getEqu(model_bmi)

Coefficients or the Slope(m) of the model is: 
0	 1730.684646399792
intercept(c) of the model is: 	 145.04962845704327
Equation:  y = 145.05+ 1730.68x^1


In [12]:
#Run a multivariate polynomial regression of the 2nd degree on all variables.

# Make predictions using the testing set
model_ploy_all = makeModelPipe(X_train,y_train)
y_pred_ploy_all = model_ploy_all.predict(X_val)
y_pred_train_ploy_all = model_ploy_all.predict(X_train)

getEqu(model_ploy_all)

Coefficients or the Slope(m) of the model is: 
0	 -294.2080086841063
1	 419.2010390675124
2	 374.02939695761484
3	 6125.604636558844
4	 -5596.239214448126
5	 -2527.789596386122
6	 142.42213197121885
7	 -1369.4860762488145
8	 35.27709489038034
9	 2738.9492546707284
10	 1978.9207071789344
11	 -531.0028700527813
12	 792.8471887140352
13	 -5281.5978425644735
14	 -4386.592852978872
15	 7198.632650422597
16	 8304.255257139772
17	 2951.63404279809
18	 322.0985115762917
19	 -1.7765698130536087
20	 663.4371695081152
21	 1903.4023322820349
22	 12428.175689038417
23	 -5472.986089482095
24	 -8259.36097606115
25	 -10103.734298958745
26	 -3504.995234990181
27	 420.6699218304134
28	 1869.469785737541
29	 3536.350127045135
30	 -15280.96178485793
31	 15611.451296621824
32	 5256.51224853989
33	 -3450.3835220897467
34	 7058.905461738664
35	 384.1981959220748
36	 -136.9551298480278
37	 28941.29049986957
38	 -22321.57708647468
39	 -12783.49677811111
40	 -1533.8713294173895
41	 -12089.878757269062
42	 -3068

In [17]:
#Compare the three models by looking at R-squared, MAPE and MAE.

def getMetrics(y,y_pred):
    print(f'Mean Absolute Precentage error: {(mean_absolute_percentage_error(y, y_pred)*100):.2f}%')
    print(f'Mean Absolute error: {mean_absolute_error(y, y_pred):.2f}')
    print(f'The model is preforming {r2_score(y,y_pred)*100:.2f}% better than a navie model')

# Comparing all the models on training data

print('For multi Linear Regression model')
getMetrics(y_train,y_pred_train_all)
print('='*_,end = '\n')

print('For Polynomial Regession model on BMI')
getMetrics(y_train,y_pred_train_bmi)
print('='*_,end = '\n')

print('For Polynomial Regession model on All data')
getMetrics(y_train,y_pred_train_ploy_all)
print('='*_,end = '\n')

For multi Linear Regression model
Mean Absolute Precentage error: 39.97%
Mean Absolute error: 43.31
The model is preforming 53.06% better than a navie model
For Polynomial Regession model on BMI
Mean Absolute Precentage error: 48.05%
Mean Absolute error: 50.84
The model is preforming 37.32% better than a navie model
For Polynomial Regession model on All data
Mean Absolute Precentage error: 34.75%
Mean Absolute error: 38.46
The model is preforming 62.25% better than a navie model


In [18]:
# Comparing all the models on Validation data

print('For multi Linear Regression model')
getMetrics(y_val,y_pred_all)
print('='*_,end = '\n')

print('For Polynomial Regession model on BMI')
getMetrics(y_val,y_pred_bmi)
print('='*_,end = '\n')

print('For Polynomial Regession model on All data')
getMetrics(y_val,y_pred_ploy_all)
print('='*_,end = '\n')

For multi Linear Regression model
Mean Absolute Precentage error: 30.31%
Mean Absolute error: 44.59
The model is preforming 47.01% better than a navie model
For Polynomial Regession model on BMI
Mean Absolute Precentage error: 38.64%
Mean Absolute error: 52.17
The model is preforming 28.33% better than a navie model
For Polynomial Regession model on All data
Mean Absolute Precentage error: 30.38%
Mean Absolute error: 46.54
The model is preforming 34.31% better than a navie model


In [19]:
# Comparing all the models on Test data

print('For multi Linear Regression model')
y_pred_test_all = model_all.predict(X_test)
getMetrics(y_test,y_pred_test_all)
print('='*_,end = '\n\n')

print('For Polynomial Regession model on BMI')
y_pred_test_bmi = model_bmi.predict(np.array(X_test['bmi']).reshape(-1,1))
getMetrics(y_test,y_pred_test_bmi)
print('='*_,end = '\n\n')

print('For Polynomial Regession model on All data')
y_pred_test_ploy_all = model_ploy_all.predict(X_test)
getMetrics(y_test,y_pred_test_ploy_all)
print('='*_,end = '\n\n')

For multi Linear Regression model
Mean Absolute Precentage error: 36.83%
Mean Absolute error: 42.91
The model is preforming 45.06% better than a navie model

For Polynomial Regession model on BMI
Mean Absolute Precentage error: 44.85%
Mean Absolute error: 52.31
The model is preforming 21.70% better than a navie model

For Polynomial Regession model on All data
Mean Absolute Precentage error: 38.47%
Mean Absolute error: 45.85
The model is preforming 35.02% better than a navie model



# Which model would you choose for deployment, and why?

-Multi Variant Linear Regression

-In comparison to the other models, it has the greatest R-squared value (0.4506), meaning that it explains more variance in the target variable.

-The Multivariate Linear Regression has the lowest MAPE (36.83), which demonstrates that, generally speaking, its predictions are the closest to the actual values.